File: Algoritimo-Estrutruras-de-Dados/Aeds-ll/2025_01/exercicios/lab_2.1/questao_02/Aquecimento.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

typedef struct{
    char nome[51];
    int peso;
} Atleta;

Atleta* NewAtleta(char* nome, int peso){
    Atleta* atleta = (Atleta*)malloc(sizeof(Atleta));
    strcpy(atleta->nome, nome);
    atleta->peso = peso;

    return atleta;
}

void DelAtleta(Atleta* atleta){
    free(atleta);
}

int Comparar(Atleta a, Atleta b){
    int result = 0; 

    if(a.peso >  b.peso){
        result = -1;
    } else if(a.peso < b.peso){
        result = 1;
    } else{
        result = strcmp(a.nome, b.nome);
    }

    return result;
}

void Swap(Atleta* a, Atleta* b){
    Atleta temp = *a;
    *a = *b;
    *b = temp;
}

void OrdenarAtletas(Atleta* atletas[], int n){
    for(int i = 0; i < n; i++){
        int max = i;
        for(int j = i + 1; j < n; j++){
            if(Comparar(*atletas[j], *atletas[max]) < 0){
                max = j;
            }
        }
        Swap(atletas[i], atletas[max]);
    }
}

int LerAtletas(Atleta* atletas[]){
    char nome[51];
    int n = 0, peso = 0;

    while(scanf(" %s %d", nome, &peso) != EOF){
        Atleta* atleta = NewAtleta(nome, peso);
        atletas[n] = atleta;
        n++;
    }

    return n; 
}

void ImprimirAtletas(Atleta* atletas[], int n){
    for(int i = 0; i < n; i++){
        printf("%s %d\n", atletas[i]->nome, atletas[i]->peso);
    }
}

void DelAtletas(Atleta* atletas[], int n){
    for(int i = 0; i < n; i++){
        DelAtleta(atletas[i]);
    }
}

int main(void){
    int n = 0;
    Atleta* atletas[100];

    n = LerAtletas(atletas);
    OrdenarAtletas(atletas, n);
    ImprimirAtletas(atletas, n);
    DelAtletas(atletas, n);

    return 0;
}

File: Algoritimo-Estrutruras-de-Dados/Aeds-ll/2025_01/exercicios/lab_2.1/questao_01/Aquecimento.java
import java.util.*;

class Atleta{
    String nome;
    int peso;

    public Atleta(){
        this.nome = "";
        this.peso = 0;
    }

    public Atleta(String nome, int peso){
        this.nome = nome;
        this.peso = peso;
    }

    public String getNome(){
        return nome;
    }

    public void setNome(String nome){
        this.nome = nome;
    }
}

public class Aquecimento{
    public static int Comparar(Atleta a1, Atleta a2){
        int result = 0;
    
        if(a1.peso > a2.peso){
            result = -1;
        } else if(a1.peso < a2.peso){
            result = 1;
        } else{
            result = a1.nome.compareTo(a2.nome);
        }
    
        return result;
    }

    public static void swap(Atleta[] atletas, int i, int j){
        Atleta temp = atletas[i];
        atletas[i] = atletas[j];
        atletas[j] = temp;
    }

    private static void OrdenarAtletas(Atleta[] atletas, int N){
        for(int i = 0; i < N; i++){
            int max = i;
            for(int j = i + 1; j < N; j++){
                if(Comparar(atletas[j], atletas[max]) < 0){
                    max = j;
                }
            }
            swap(atletas, i, max);
        }
    }

    private static void ImprimirAtletas(Atleta[] atletas){
        for(Atleta atleta : atletas){
            System.out.println(atleta.nome + " " + atleta.peso);
        }
    }

    private static void LerAtletas(Scanner sc, Atleta[] atletas, int N){
        for(int i = 0; i < N; i++){
            String nome = sc.next();
            int peso = sc.nextInt();
            atletas[i] = new Atleta(nome, peso);
        }
    }

    public static void main(String[] args){
        Scanner sc = new Scanner(System.in);
        int N = sc.nextInt();
        Atleta[] atletas = new Atleta[N];

        LerAtletas(sc, atletas, N);
        OrdenarAtletas(atletas, N);
        ImprimirAtletas(atletas);

        sc.close();
    }
}
